### PR TITLE
Alternative structure for type-assigning client options.

### PR DESCRIPTION
**Pull request by reggi**
This is a prototype for an alternative way of handling client options in the driver. it loops over options and uses a lookup table to designate type. The lookup table itself is used to cast what the URI options looks like as a type.

Major difference between this and other prototypes is there is no option validation happening here. It's a mechanism for assigning uri string types to what their type should be.

### CODE REVIEW

File: src/options.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    const { type, normalized } = ref;
+    let currentValue = options[normalized as keyof typeof MongoURIOptions];
+    let results;
+
+    function recursive(
+      normalized: string,
+      value: string | string[] | undefined,
+      type: URIType | Lookup
+    ): URIValue<typeof MongoURIOptions> {
+      const match = (t: URIType) => type === t;
```
Comment (nbbeeken) [inline src/options.ts:181]: Very ML approach here with the match (which is cool), is there advantage over a switch block?

Comment (reggi) [reply src/options.ts:181]: I think I did originally do a switch, and I switched it (pun-intended) I'm a little switch-adverse, and I know Mr. Broadstone is as well.

File: src/cmap/auth/defaultAuthProviders.ts [hunk 2]
```diff
@@ -7,24 +7,26 @@ import { MongoDBAWS } from './mongodb_aws';
...
 export enum AuthMechanism {
-  MONGODB_AWS = 'MONGODB-AWS',
-  MONGODB_CR = 'MONGODB-CR',
-  MONGODB_DEFAULT = 'DEFAULT',
-  MONGODB_GSSAPI = 'GSSAPI',
-  MONGODB_PLAIN = 'PLAIN',
-  MONGODB_SCRAM_SHA1 = 'SCRAM-SHA-1',
-  MONGODB_SCRAM_SHA256 = 'SCRAM-SHA-256',
-  MONGODB_X509 = 'MONGODB-X509'
+  'MONGODB-AWS' = 'MONGODB-AWS',
```
[the remark this answers is not in the document]
Comment (reggi) [reply src/cmap/auth/defaultAuthProviders.ts:10]: Yeah, for options we can support enums but we should allow strings. We can keep the enum as it was before this change, and create another type for the strings?

File: src/mongo_client.ts [hunk 3]
```diff
@@ -102,7 +104,7 @@ export interface MongoURIOptions extends Pick<WriteConcernOptions, 'journal' | '
   /** Specify the database name associated with the user’s credentials. */
   authSource?: string;
   /** Specify the authentication mechanism that MongoDB will use to authenticate the connection. */
-  authMechanism?: AuthMechanism;
+  authMechanism?: AuthMechanisms;
```
[the remark this answers is not in the document]
Comment (reggi) [reply src/mongo_client.ts:107]: Yeah this goes for all the enums that are user-enterable. We can't expect our users to import the enum and pass it in.

File: src/mongo_client.ts [hunk 4]
```diff
@@ -205,8 +207,13 @@ export interface MongoClientOptions
   driverInfo?: DriverInfo;
   /** String containing the server name requested via TLS SNI. */
   servername?: string;
-
   dbName?: string;
+  auto_reconnect?: boolean;
```
[the remark this answers is not in the document]
Comment (reggi) [reply src/mongo_client.ts:211]: Gotcha, but would we want to notify this change?

File: src/read_concern.ts [hunk 5]
```diff
@@ -6,14 +6,16 @@ export enum ReadConcernLevel {
...
+
 /**
  * The MongoDB ReadConcern, which allows for control of the consistency and isolation properties
  * of the data read from replica sets and replica set shards.
  *
  * @see https://docs.mongodb.com/manual/reference/read-concern/index.html
  */
 export class ReadConcern {
-  level: ReadConcernLevel;
+  level: ReadConcernLevels;
```
[the remark this answers is not in the document]
Comment (reggi) [reply src/read_concern.ts:18]: Makes sense, but it just means the `readConcern` client option is more complicated and needs to be properly handled / resolved.

```
interface Client {
  readConcern: ReadConcern | { level: ReadConcernLevel | ReadConcernLevels }
}
```

**Pull request closed without merging**